azurerm/resource_arm_front_door.go

```diff
@@ -0,0 +1,1291 @@
+										Required:     true,
+										ValidateFunc: validation.IntBetween(1, 5),
+									},
+									"host_header": {
+										Type:     schema.TypeString,
+										Optional: true,
+									},
+								},
+							},
+						},
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: azure.ValidateBackendPoolRoutingRuleName,
+						},
+						"health_probe_name": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"load_balancing_name": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"frontend_endpoint": {
+				Type:     schema.TypeList,
+				MaxItems: 100,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: azure.ValidateBackendPoolRoutingRuleName,
+						},
+						"host_name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"session_affinity_enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  true,
+						},
+						"session_affinity_ttl_seconds": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"custom_https_configuration": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"certificate_source": {
+										Type:     schema.TypeString,
+										Optional: true,
+										ValidateFunc: validation.StringInSlice([]string{
+											string(frontdoor.CertificateSourceAzureKeyVault),
+											string(frontdoor.CertificateSourceFrontDoor),
+										}, false),
+										Default: string(frontdoor.CertificateSourceFrontDoor),
+									},
+									// NOTE: None of these attributes are valid if
+									//       certificate_source is set to FrontDoor
+									"azure_key_vault_certificate_secret_name": {
+										Type:     schema.TypeString,
+										Optional: true,
+									},
+									"azure_key_vault_certificate_secret_version": {
+										Type:     schema.TypeString,
+										Optional: true,
+									},
+									"azure_key_vault_certificate_vault_id": {
+										Type:     schema.TypeString,
+										Optional: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+
+			"friendly_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceArmFrontDoorCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	if err := azure.ValidateFrontdoor(d); err != nil {
```

Comment:
We could do this on the object we pass to the api, thus sharing the `d.Get` logic?